# Crosslinking IMS

This application attempts to find all cross-linked peptides in the given Ion Mobility MS dataset. 
First, the FASTA file is read in and possible peptides are generated by digesting the protein(s) 
and assuming 1 max missed cleavage. The possible peptides are searched for in the given features 
by matching up the monoisotopic mass of the feature to the mass of the peptides. The mass ppm 
tolerance is an input parameter of the application. All possible cross-links of the 
unmodified peptides are then calculated and searched for in the Isotopic Peak data. 
A detailed csv results file is created at the end.

## Syntax

```
CrossLinkingIMSConsole.exe 
  -f [Features File] 
  -p [Peaks File] 
  -fasta [FASTA file] 
  -ppm Value 
  [optional arguments]
```

###  Required Arguments

| Argument | Description |
|----------|-------------|
| -f       | Features File. LC-IMS-MS Feature Finder Output|
| -p       | Peaks File. DeconTools Output|
| -fasta   | FastA File. Contains all protein sequences to search.|
| -ppm [value] |  Mass tolerance in ppm |

### Optional Arguments

| Argument | Description |
|----------|-------------|
| -c       | The maximum number of missed cleavages to consider. Defaults to 1. |
| -t       | Set to 'full' for fully tryptic only.  Set to 'partial' to consider partially and fully tryptic. Set to 'none' to consider non, partially, and fully tryptic. Defaults to 'full'. |
| -o       |The desired location and name for the output file. Defaults to workingDirectory/crossLinkResults.csv |

Crosslinked peptide pairs are found using a delta mass value that is
typically computed based on the number of carbons and number of nitrogen atoms,
treating carbons in the crosslinked peptide as C13 and nitrogens as N15.
Alter this behavior using switches -C13off, -N15off, and -Static

| Argument         | Description |
|------------------|-------------|
| -C13off          | Peptides are not labeled with C13 |
| -N15off          | Peptides are not labeled with N15 |
| -Static [Value]  | Static (fixed) mass shift to use for all paired peptides |

For example, to disable C13 and N15 shifts and only use a
fixed shift of 25.5 Da, your command line would look like this (all on one line):

```
CrossLinkingIMSConsole.exe
 -f Dataset_LCMSFeatures.txt 
 -p Dataset_Filtered_isos.csv 
 -fasta Database.fasta
 -ppm 10 -C13Off -N15off –Static 25.5
```

## Features File

This is a tab-delimited file normally generated by the [LC-IMS-MS Feature Finder](https://github.com/PNNL-Comp-Mass-Spec/LC-IMS-MS-Feature-Finder), 
which is in-house software developed at Pacific Northwest National Lab. If you need a copy of the software, see 
[https://omics.pnl.gov/software](https://omics.pnl.gov/software/LC-IMS-MS_Feature_Finder.php)

A normal Features file is created by clustering together isotopic profiles that are believed to be 
from the same isotopic signature eluting over a period of time. Each feature in the file has 
representative mass, elution time, drift time and charge state values that define the feature.

To fake a Features file, you can create a tab-delimited text file with the following columns required by the CrossLinkingIMS software: 
| Column        | Description |
|---------------|-------------|
| Feature_Index | Integer identifier for the feature. Any numbers can be used, but it is suggested to use sequential number. |
| Monoisotopic_Mass | The monoisotopic mass of the LC-IMS-MS Feature. |
| Scan_Start    | The first LC Scan where the Feature was observed. |
| Scan_End      | The last LC Scan where the Feature was observed. |
| Scan          | The LC Scan where the Feature was most abundant. |
| IMS_Scan      | IMS Scan of the most abundant LC Scan where the Feature was most abundant. |
| Drift_Time    | The detected drift time of the Feature. |
| Abundance     |  The total intensity of the Feature. |


## Peaks File

This is a tab-delimited file that can be exported when running [DeconTools](https://github.com/PNNL-Comp-Mass-Spec/DeconTools/)
If you need a copy of that software, visit either [GitHub](https://github.com/PNNL-Comp-Mass-Spec/DeconTools/releases)
or [https://omics.pnl.gov/software](https://omics.pnl.gov/software/decontools-decon2ls)

A peaks file can easily be manually generated since it is simply a dump of all Isotopic Peaks in the data.

To generate a peaks file, create a tab-delimited text file with the following columns required by the CrossLinkingIMS software:

| Column     | Description |
|------------|-------------|
| frame_num  | The LC Scan of the peak. |
| scan_num   | The IMS Scan of the peak. |
| mz         | The m/z value of the peak. |
| intensity  | The intensity of the peak. |


## Contacts

Written by Kevin Crowell for the Department of Energy (PNNL, Richland, WA) \
E-mail: proteomics@pnnl.gov \
Website: https://panomics.pnl.gov/ or https://omics.pnl.gov

## License

The Crosslinking IMS software is licensed under the Apache License, Version 2.0; 
you may not use this file except in compliance with the License.  You may obtain 
a copy of the License at https://opensource.org/licenses/Apache-2.0
